HPCC-16375 Add core functionality

AttilaVamos opened the pull request:
Add core functionality to copy and extend file history in DFU
operation like: spray, copy.

It is the first subtask of "HPCC-16072 retain metadata (atleast
the owner) during file copy" issue.

Signed-off-by: Attila Vamos attila.vamos@gmail.com

AttilaVamos commented:
@jakesmith This is the core functionality.  Please review.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3cad697d882a14fb55afbcaa5c9d1230505cb491
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
Not sure if some comments are getting lost, or not seen...
e.g. : https://github.com/hpcc-systems/HPCC-Platform/pull/9108/files#r80935677

@AttilaVamos - can you comment on each of the previous review comments please. Can comment in previous PR if easier.

AttilaVamos commented:
@jakesmith done.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3cad697d882a14fb55afbcaa5c9d1230505cb491
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[32m [   OK    ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[33m [ TIMEOUT ] \x1b[0m 
HPCC status:
mydafilesrv     ( pid    17061 ) is running ...
mydali          ( pid    17214 ) is running ...
mydfuserver     ( pid    17383 ) is running ...
myeclagent      ( pid    17559 ) is running ...
myeclccserver   ( pid    17721 ) is running ...
myeclscheduler  ( pid    17891 ) is running ...
myesp           ( pid    18057 ) is running ...
myroxie         ( pid    18258 ) is running ...
mysasha         ( pid    18477 ) is running ...
mythor          is stopped,
mythor          is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9197
~/smoketest/PR-9197 ~/smoketest/PR-9197/build
./unittest.sh
~/smoketest/PR-9197/build
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3cad697d882a14fb55afbcaa5c9d1230505cb491
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: Fail
HPCC start:
Starting mydafilesrv ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting mydali ...           \x1b[32m [   OK    ]  \x1b[0m 
Starting mydfuserver ...      \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclagent ...       \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclccserver ...    \x1b[32m [   OK    ]  \x1b[0m 
Starting myeclscheduler ...   \x1b[32m [   OK    ]  \x1b[0m 
Starting myesp ...            \x1b[32m [   OK    ]  \x1b[0m 
Starting myroxie ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mysasha ...          \x1b[32m [   OK    ]  \x1b[0m 
Starting mythor ...           \x1b[33m [ TIMEOUT ] \x1b[0m 
HPCC status:
mydafilesrv     ( pid    20738 ) is running ...
mydali          ( pid    20898 ) is running ...
mydfuserver     ( pid    21069 ) is running ...
myeclagent      ( pid    21242 ) is running ...
myeclccserver   ( pid    21409 ) is running ...
myeclscheduler  ( pid    21578 ) is running ...
myesp           ( pid    21744 ) is running ...
myroxie         ( pid    21945 ) is running ...
mysasha         ( pid    22163 ) is running ...
mythor          is stopped,
mythor          is stopped,
pushd /mnt/disk1/home/vamosax/smoketest/PR-9197
~/smoketest/PR-9197 ~/smoketest/PR-9197/build
./unittest.sh
~/smoketest/PR-9197/build
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest 
Sha: 3cad697d882a14fb55afbcaa5c9d1230505cb491
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+            // Handle history
+            IPropertyTree *curHistory = lock.queryHistory();
+
+            if (srcHistory)
```
Can this not exist?

AttilaVamos replied:
I think from now it always exists, but can be empty.

jakesmith replied:
Was true, but it will always exist with queryHistory() change I think.

AttilaVamos replied:
Yes, it is.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+            }
+
+            // Add new record about this operation
+            IPropertyTree * newRecord = createPTree("Origin");
```
Safer to make it an Owned and use getClear() when you transfer ownership, i.e. when you do the addPropTree call.

jakesmith replied:
minor: You don't actually need to name the node here (and similar in a couple of other places), doing a add/setPropTree with the name is sufficient.

AttilaVamos replied:
Ok, I will update.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+
+            StringBuffer val;
+            srcAttr->getProp("@owner", val);
+            newRecord->addProp("@owner", val.str());
```
no need to temporary 'val', use addProp(.... , queryProp..)

AttilaVamos replied:
Ok, I will update

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+
+            srcAttr->getProp("@workunit", val.clear());
+            if (0 == val.length())
+                val.set("N/A");
```
Why is this better than having no @workunit ?

AttilaVamos replied:
If something not there which usually should be then I can feel there is a bug somewhere. So in case of Spray there is no previous workunit and originally it filled with and empty string, which can be suspicious as well. That's why I store that "N/A" string.
Moreover if some filed missing, than it can cause further headache when the history would be export as CSV.

jakesmith replied:
If the file was not created by a workunit, it will not have a @workunit attribute.

There's no need to hard-wire "N/A" into serialized published meta data for false case.
You can interpret that at read time.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+            newRecord->addProp("@operation", getOperationTypeString());
+
+            // add original file name
+            if (distributedSource)
```
null when dealing with multiple source files? 
Could do with a comment if so.

AttilaVamos replied:
Yes it is null in Spray case. I will put an explanation comment there.

jakesmith replied:
Ok

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
 }
 
+void FileSprayer::splitAndStoreFileInfo(IPropertyTree * newRecord, RemoteFilename &remoteFileName
+                                      , aindex_t idx, bool isDistributedSource)
```
trivial: slightly strange spacing. Would normally have this leading comma as trailing comma on 1st line.

AttilaVamos replied:
OK, I will change.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+    remoteFileName.split(&drive, &path, &fileName, &ext);
+    if (idx == 0)
+    {
+        if(drive.isEmpty())
```
formatting: space after if

AttilaVamos replied:
Ok, I will change

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+
+        newRecord->addProp("@path", path.str());
+    }
+    if (!isDistributedSource && ext.length())
```
The !isDistributedSource is to do with multiple source file case?
Needs some comments to explain what and why it is conditionally adding the extension here.

jakesmith replied:
See other comment relating to where splitAndStoreFileInfo.
The common part (idx == 0) block, and the 2 different parts for each use case (single vs multiple sources) - would be clearer if separated.

AttilaVamos replied:
I will put explanation and when we can agree about splitAndStoreFileInfo() function.

jakesmith replied:
Ok

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+                {
+                    FilePartInfo & curSource = sources.item(idx);
+                    RemoteFilename remoteFile = curSource.filename;
+                    splitAndStoreFileInfo(newRecord, remoteFile, idx, false);
```
I think the mixed used of splitAndStoreFileInfo for single and multi source handling is a bit confusing.
Handling the 1st part (0 == idx) to handle the drive/path/ip extraction vs others... followed by different code for each case.

It would be clearer if you did that 1st unconditionally (before line 3298) as both these branch have to do.
Then dealt with each case separately, as code is not the same.

AttilaVamos replied:
Originally that code was two places in true and false branch (as you mentioned in this comment) but I heard in my mind you comment about similar code should summed. 
And the other reason to create this function was: both code worked on a RemoteFileName object and doing same separation and assignments.

jakesmith replied:
If truly common duplicated code or code that can be parametrized then it makes sense to move to a shared function.
I can see why you've commoned up now, because both paths take a RemoteFilename from different origins.

Ok, well I think worth adding a couple of comments re. handling of single source vs multi source

AttilaVamos replied:
Ok. I will add explanations

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+    operation = op;
+}
+
+dfu_operation FileSprayer::getOperation(void)
```
style: we don't tend to prototype methods as (void).

Also could be const method.

AttilaVamos replied:
Ok, I will change.

jakesmith reviewed (changes requested):
@AttilaVamos please see comments.

AttilaVamos commented on dali/base/dadfs.cpp:
```diff
@@ -3472,6 +3485,7 @@ protected: friend class CDistributedFilePart;
             queryAttributes().setPropInt64("@size", totalsize);
         if (useableCheckSum)
             queryAttributes().setPropInt64("@checkSum", checkSum);
+        setHistory(fdesc);
```
Ok

AttilaVamos replied:
Yes it means History always present from now. If it is not then I can go back the square one with reference to a non existent object.

jakesmith replied:
Either it should always create (and you do) and return a reference so interface is consistent.
Or it should not create blank History nodes, which are not used the vast majority of the time and end up being serialized and committed to meta store.

It was I thought going in the direction of not insisting it was there to avoid waste, therefore returning it optionally made sense.
If you are going to force it's creation, and then the only case it's not there is legacy, it doesn't make sense to return a pointer which will be null only for legacy. It should be consistent and if create most of time, create all the time.

It's not a big deal to have the extra empty usually unused History node serialize/part of meta store, so leave it like it is, but make sure always exists (will only not in legacy), add a comment to say why creating in that case (in queryHistory()).

AttilaVamos replied:
Ok.

AttilaVamos commented on dali/base/dadfs.hpp:
```diff
@@ -494,6 +496,10 @@ class DistributedFilePropertyLock {
+    IPropertyTree *queryHistory()
+    {
+        return file->queryHistory();
+    }
```
Ok

AttilaVamos commented on dali/base/dadfs.hpp:
```diff
@@ -457,7 +459,7 @@ interface ISimpleSuperFileEnquiry: extends IInterface // lightweight local
 /*
  * Context-based file property locking mechanism. Allows early unlocking for special cases,
- * stores the reload flag and allows you to query the 'Attr' section of the file property,
+ * stores the reload flag and allows you to query the 'Attr' and 'History' sections of the file property,
```
Ok

AttilaVamos commented on dali/base/dadfs.cpp:
```diff
@@ -3512,6 +3526,15 @@ protected: friend class CDistributedFilePart;
+        if (isEmptyPTree(history))
+            resetHistory();
+        else
+            resetHistory(createPTreeFromIPT(history));
```
Originally when everything was in one PR (core and dfuplus related) then was a createHistory() and eraseHistory() and you mentioned they are same functionality.
Now the two implementation is separated and I created this method which is used here to create  History  will be used in DFUPlus eraseHistory as well.

AttilaVamos commented on dali/base/dafdesc.cpp:
```diff
@@ -1340,17 +1352,31 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
         }
-    }
 
+        const char * historyStr = "History";
```
OK, What kind of history? In a byte or int variable? What kind of versioning, sequence number, bit field?

jakesmith replied:
I think I'd use 2 bytes I think (unsigned short)
Define for current version static const unsigned short serializationVersion = 1;
and add a comment above the 2 existing #define's to say they are legacy and cannot be changed.

AttilaVamos replied:
OK.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
+                ForEachItemIn(idx, sources)
+                {
+                    FilePartInfo & curSource = sources.item(idx);
+                    RemoteFilename remoteFile = curSource.filename;
```
Looks like should be a reference

AttilaVamos replied:
Ok

AttilaVamos commented:
@jakesmith it is updated, built, started but not tested yet. Please re-review.

HPCCSmoketest commented:
Automated Smoketest 
Sha: b7f1b7b43c6f4063e0882a97624ab024f7ccd10c
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -1353,14 +1358,26 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
-        const char * historyStr = "History";
-        if (mb.remaining() > strlen(historyStr))
+        // If we have more serialised data in MemoryBuffer
+        if (mb.remaining() > sizeof(extendedVersion))
```
minor: Should that be >= ?
In reality for this version it will always be >, but..

AttilaVamos replied:
Updated

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -1353,14 +1358,26 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
+            {
+                //If extendedVersion >= 1 then we should have stored History
+                const char * historyStr = "History";
+                if (mb.remaining() > strlen(historyStr))
```
It's now an error if it is not what's expected, so should fire exceptions just as you do if version=0 if anything doesn't match. In fact you should assume it does, just as everything else does.
You've read version it's >= MIN , now deserialize ptree without further validation.

And to make that crystal clear, the block could start with:
```
if (extendedVersion >= VERSION_MINHISTORY)
```

AttilaVamos replied:
Updated

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -37,9 +37,13 @@
 
 #define SDS_CONNECT_TIMEOUT  (1000*60*60*2)     // better than infinite
 
+// They are legacy and cannot be changed.
```
They -> These

AttilaVamos replied:
Updated

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -1451,6 +1468,9 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
         else
             attr.setown(createPTree("Attr"));
+
+        extendedVersion = serializationVersion;
```
Normally you'd use the serialization version to validate during deserialization only, you wouldn't need or store it beyond that, so I'm not sure it needs to be a class member like this.

AttilaVamos replied:
Updated

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -937,6 +941,7 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
     SocketEndpointArray *pending;   // for constructing cluster group
     bool setupdone;
     byte version;
+    unsigned short extendedVersion = 0;
```
Normally you'd use the serialization version to validate during deserialization only, you wouldn't need or store it beyond that, so I'm not sure it needs to be a class member like this.

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3258,21 +3254,21 @@ void FileSprayer::updateTargetProperties()
+            IPropertyTree &curHistory = lock.queryFile()->queryHistory();
 
-            if (srcHistory)
+            //if (srcHistory)
```
Should remove if satisfied never null

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3284,42 +3280,42 @@ void FileSprayer::updateTargetProperties()
             StringBuffer val;
-            srcAttr->getProp("@owner", val);
-            newRecord->addProp("@owner", val.str());
+            //srcAttr->getProp("@owner", val);
```
should remove

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3284,42 +3280,42 @@ void FileSprayer::updateTargetProperties()
             newRecord->addProp("@timestamp",temp.getString(timestr).str());
 
             StringBuffer val;
```
unused (except incorrectly below on addProp("@workunit" line)

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3284,42 +3280,42 @@ void FileSprayer::updateTargetProperties()
-                val.set("N/A");
-            newRecord->addProp("@workunit", val.str());
+            if (srcAttr->hasProp("@workunit"))
+                newRecord->addProp("@workunit", val.str());
```
val is not set.
This should suffice:
```
newRecord->setProp("@workunit", srcAttr->queryProp("@workunit"));
```

jakesmith replied:
btw addProp works for attributes (@) but setProp is more correct, since you can't have multiple attributes of the same name.

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3284,42 +3280,42 @@ void FileSprayer::updateTargetProperties()
             {
+                // add original file name from a single source
                 RemoteFilename remoteFile;
                 distributedSource->getPart(0)->getFilename(remoteFile, 0 );
```
formatting: extra space before );

jakesmith replied:
This leaks.

should use queryPart

AttilaVamos replied:
Updated

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3284,42 +3280,42 @@ void FileSprayer::updateTargetProperties()
                 }
             }
-            curHistory->addPropTree("Origin",newRecord);
+            curHistory.addPropTree("Origin",newRecord);
```
Leak.
Should use newRecord.getClear()

AttilaVamos replied:
Updated

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2738,6 +2738,11 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
 
+    IPropertyTree &queryHistory() const
+    {
+        return *root->queryPropTree("History");
```
Previously discussed (https://github.com/hpcc-systems/HPCC-Platform/pull/9197#discussion_r81319781)
If you're forcing History node to exist and returning a reference.
You need to check here too and create if missing (will only apply to legacy) - and comment as to why.

AttilaVamos commented:
@jakesmith updtated. Please check it again.

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -37,9 +37,14 @@
 
+#define VERSION_MINHISTORY 1  // Add History to serialisation/de-serialisation
+// This version serialised in the extendedVersion of CFileDescriptor
+static const unsigned short serializationVersion = VERSION_MINHISTORY;
```
trivial: but you expect the version to increment as time goes on - it's a coincidence it's equal to the min version of the last feature added, so looks little odd to say this.
Better to just say = 1 here.

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -1353,14 +1357,29 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
+                    if (nullptr != strstr(buffer, historyStr))
+                        history.setown(createPTree(mb));
+                    else
+                        history.setown(createPTree(historyStr));
```
See previous comment : https://github.com/hpcc-systems/HPCC-Platform/pull/9197#discussion_r81339822

jakesmith replied:
i.e. the version tells you have history, just deserialize it. i.e. just this line not strstr etc..

AttilaVamos replied:
That comment link lead me somewhere strange, seems irrelevant place.

jakesmith replied:
Link works for me.
Remade the point above anyway.

You shouldn't be manually checking (strstr etc), just deserialize it. You have already validated the version. So:
```
if (extendedVersion >= VERSION_MINHISTORY)
    history.setown(createPTree(historyStr));
else
    throw ...
```

AttilaVamos replied:
Your suggestion is wrong. That code fragment doesn't deserialise the history, but create a new and empty one. 

My code deseirialise history if there is a valid desriailisable one (contains valid characters to create xpath)  or create a new and empty one if not.
And what happened if the memory buffer destroyed/altered (after it serialised, trough the transport, etc) somehow (e.g. overwritten, truncated)? I know because I experienced it. Crash/exception somewhere other place (in the tree creation). Should we continue to process if we (can) know there is a missing bit and it will be unsuccessful?

So, I can throw an exception if there is not the expected (vital) string in in the mb from current position, like this:
```
const char * buffer = mb.toByteArray()+mb.getPos();
 if (nullptr != strstr(buffer, historyStr))
     history.setown(createPTree(mb));    // There is a valid History in mb deserialise it
else
    throw MakeStringException(-1,"Missing History from the extended version of FileDescriptor!");
```
or I try to catch it with:
```
try
     history.setown(createPTree(mb));    // Try to deserialise the content of the mb
catch(...)
...
    throw MakeStringException(-1,"Missing History from the extended version of FileDescriptor!");
```
Either way I think I should handle/prepare for the problem (a necessary paranoid behaviour to survive).
However I will implement what you want, with or without checking, because without your acceptance this code won't be merged.
In this case the code would be this:
```
 if (extendedVersion >= VERSION_MINHISTORY)
    history.setown(createPTree(mb));    // Deserialise the content of the mb
    // Further deserialisations should place here
 else
   throw MakeStringException(-1,"FileDescriptor extended serialization version shouldn't be zero!");
``

jakesmith replied:
>Your suggestion is wrong. That code fragment doesn't deserialise the history, but create a new and empty one.

That was a typo'. I should of course have read:
```

You shouldn't be manually checking (strstr etc), just deserialize it. You have already validated the version. So:

if (extendedVersion >= VERSION_MINHISTORY)
    history.setown(createPTree(mb));
else
    throw ...
```

I can see no good reason to do further [partial] validation of the content of mb, with a strlen and a strstr call, beyond the version check. No more so than for any other part of deserializing the file descriptor or for that matter in any other serialize/deserialize code elsewhere in the system.

Sure it could be corrupted, truncated, etc. etc. - but if there were issues like that, then it could hit deserialize problems at any stage of the the deserializtion of the file descriptor, History isn't special in that sense.
It doesn't make sense to add this kind of manual partial checking of the content of serialized data ahead of deserializing it everywhere.

Now, you _could_ argue that validation and error handling *in* the deserialize methods should be more robust, but that's a separate and general issue.

After successful deserialization you _could_ add a assertex(streq("History", history->queryName());

The same applies to try/catch, you could, but unexpected result could happen during any part of the deserialization, not just History and it would be impossible to recover satisfactorily in most cases, nor would it be desirable in most cases to continue.

@AttilaVamos - if you have some concerns as to why History is more likely to be corrupted during serialization/deserialization than any other part of the serialization/deserialization, then.. maybe I've missed something. And if you really think it's important that these partial memory buffer checks remain ahead of deserializing it for History specifically then leave them in and add a comment in the code as to why these additional checks are done here for History.

AttilaVamos replied:
All right. Over-paranoid checking removed.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 5fd3774baced461e83351807e0da47df122126d9
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2738,6 +2738,15 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+        IPropertyTree *history = root->queryPropTree("History");
+        //It can be null if the file created with an older version
+        if (!history)
+            history = root->setPropTree("History",createPTree("History")); // takes ownership
```
does this method compile ? Since it's const but altering member (root).

AttilaVamos replied:
Yes, it compile, like the "Attr" related few lines above.

jakesmith replied:
Ok

jakesmith replied:
It's not like the queryAttributes() above..
That is not a const method..

jakesmith replied:
TBH - I can't remember the semantics.. I thought if it was calling a non-const method of a member object within a const method (as this is) then it was an error.

AttilaVamos replied:
From  Bjarne Stroustrup The C++ Programming Language, 4th Edition (Clasese, page 464):
"Note that const does not apply (transitively) to objects accessed through pointers or references.
The human reader may consider such an object as ‘‘a kind of subobject,’’ but the compiler does not know such pointers or references to be any different from any others. That is, a member pointer does not have any special semantics that distinguish it from other pointers."

jakesmith replied:
So legal, but logically this object (root) is a member of class and this method is altering it.
Anyway, interesting more than anything, I wouldn't change, I'm sure there are plenty of other examples in our code base.
Also interestingly looks like they may introduce a mechanism to protect member pointer objects : http://en.cppreference.com/w/cpp/experimental/propagate_const

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -1340,17 +1356,46 @@ class CFileDescriptor:  public CFileDescriptorBase, implements ISuperFileDescrip
 
+            if (extendedVersion >= VERSION_MINHISTORY)
+            {
+                //If extendedVersion >= 1 then we should have stored History
```
superfluous comment (the line above is now self explanatory and more accurate)

AttilaVamos replied:
Removed.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3252,112 @@ void FileSprayer::updateTargetProperties()
+            if (srcAttr->hasProp("@workunit"))
+                newRecord->setProp("@workunit", srcAttr->queryProp("@workunit"));
+
+            newRecord->addProp("@operation", getOperationTypeString());
```
minor: should be setProp really

AttilaVamos replied:
and all others as well?

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3252,112 @@ void FileSprayer::updateTargetProperties()
+            StringBuffer timestr;
+            newRecord->addProp("@timestamp",temp.getString(timestr).str());
+
+            newRecord->addProp("@owner", srcAttr->queryProp("@owner"));
```
As prev. mentioned attributes can't have multiple values, so setProp is more appropriate.

AttilaVamos replied:
Changed

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -3249,13 +3252,112 @@ void FileSprayer::updateTargetProperties()
+            temp.setTime(hour, min, sec, 0);
+
+            StringBuffer timestr;
+            newRecord->addProp("@timestamp",temp.getString(timestr).str());
```
setProp

AttilaVamos replied:
Changed all relevant addProp() to setProp()

AttilaVamos commented:
@jakesmith update pushed.

HPCCSmoketest commented:
Automated Smoketest 
Sha: fcec9491d0ac0735badcc811ab05cc8bf33cd009
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
@AttilaVamos - one remaining issue I think.

AttilaVamos commented:
@jakesmith Your question is answered.

jakesmith commented:
@AttilaVamos - the issue remaining I was referring to was : https://github.com/hpcc-systems/HPCC-Platform/pull/9197/files#r81365888

AttilaVamos commented:
@jakesmith Your question is answered.

AttilaVamos commented:
@jakesmith updated, please check it again.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 41b6f1f9cd61909c2d552784bdc4d64dc0baecda
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented:
Thanks, looks good.
Can you squash the 2 commits, then ready to merge.
@AttilaVamos

AttilaVamos committed:
HPCC-16375 Add core functionality

Add core functionality to copy and extend file history in DFU
operation like: spray, copy.

It is the first subtask of "HPCC-16072 retain metadata (atleast
the owner) during file copy" issue.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith I will do it, but I should test back-/forward compatibility and between systems first.

jakesmith commented:
@AttilaVamos - ok.

AttilaVamos commented:
@jakesmith the upward compatibility is broken at the point where my paranoid code situated. The memory buffer comes from previous version contains extra (non-zero) bytes after the real content.

jakesmith commented:
It's broken because the getFileTree serialization is horrible hacked.
It serializes the file tree, then conditionally serializes extra data.
Both the request and the response are a mess and already have a couple of kludges in them, that make extending the file descriptor serialization hard if not impossible without breaking compatibility. 

The problem here should be documented in the code explaining what will break and why if the serialization format is changed/extended.

However, there is an alternative instead of extending the seriazliation format itself, the new History tree could be part of the existing Attr file descriptor node.
That won't change the serialization/deserialize code itself and should not I think cause any compatibility problems.
@AttilaVamos

jakesmith commented:
@AttilaVamos - unrelated to above, Attila has also raised the concern that History in general will increase the serialization footprint and cost inefficiency for components that don't care about History.

I'm not sure it's a big concern, but worth thinking about in some cases.
The biggest case where it would be a pity to serialize History is when Thor serializes a single part(s) to slaves and it does not care/want them.
It would be nice to have a flag to exclude History from the serialization.

But I'd get this core functionality done in this PR and then worry about this optimization.

AttilaVamos committed:
Re-work.
Store History inside the Attr

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith reworked version, please review.

HPCCSmoketest commented:
Automated Smoketest 
Sha: 741ea76072322254555bca1c29f18770aaf98349
Build: success
Install hpccsystems-platform-community_6.1.0-trunk0.el6.x86_64.rpm
HPCC Start: OK
Unittest result: total:83 passed:83 failed:0

HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2738,6 +2738,14 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+    {
+        IPropertyTree *attr =  root->queryPropTree("Attr");
+        if (attr)
+            return attr->queryPropTree("History");
```
trivial: could have been : 
```
    return root->queryPropTree("Attr/History");
```

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2782,6 +2790,14 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
         root->removeProp("Attr");
         return NULL;
     }
+    IPropertyTree * resetHistory(IPropertyTree *history=NULL)
```
What purpose does the parameter serve? Not used below other than to set to blank vs delete.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2782,6 +2790,14 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+        if (!history)
+            return queryAttributes().setPropTree("History",createPTree("History"));
+        else
+            queryAttributes().removeTree("History");
```
removeTree takes a IPT
Pity compiler doesn't spot.

jakesmith commented on dali/base/dadfs.cpp:
```diff
@@ -2782,6 +2790,14 @@ class CDistributedFileBase : implements INTERFACE, public CInterface
+        else
+            queryAttributes().removeTree("History");
+        return nullptr;
+    }
```
A few things odd/wrong with this resetHistory function.
passes in a param, which could be boolean, which either deletes history and replaces with empty node, or just deletes history.
Then returns either the new empty tree, or null.
resetTree takes a IPT but is being passed a string.

Shouldn't this just be:
```
void resetHistory()
{
    queryAttributes().removeProp("History");
}
```

?

jakesmith commented on dali/ft/filecopy.hpp:
```diff
@@ -129,6 +151,10 @@ interface IFileSprayer : public IInterface
+    virtual void setOperation(dfu_operation op) = 0;
+    virtual dfu_operation getOperation() const = 0;
+    virtual const char * getOperationTypeString() const = 0;
+
```
formatting: extra newline

jakesmith commented:
@AttilaVamos - some issues with resetHistory, but otherwise looks good.

jakesmith commented:
@AttilaVamos - I don't foresee any forward/backward compatibility issues, but have you re-tested ?

jakesmith commented:
@AttilaVamos - as resetHistory isn't used yet, you can fix it in the next subissue PR that uses it.
